sp-sim/src/sensors.rs

```diff
@@ -0,0 +1,227 @@
+                    SensorReading {
+                        value: Err(error.value),
+                        timestamp: error.timestamp,
+                    }
+                }
+            }
+            Self { last_data: Some(data), last_error: None } => SensorReading {
+                value: Ok(data.value),
+                timestamp: data.timestamp,
+            },
+            Self { last_data: None, last_error: Some(error) } => {
+                SensorReading {
+                    value: Err(error.value),
+                    timestamp: error.timestamp,
+                }
+            }
+            Self { last_data: None, last_error: None } => SensorReading {
+                value: Err(SensorDataMissing::DeviceNotPresent),
+                timestamp: 0, // TODO(eliza): what do?
+            },
+        }
+    }
+}
+
+impl Sensors {
+    pub(crate) fn from_component_configs<'a>(
+        cfgs: impl IntoIterator<Item = &'a SpComponentConfig>,
+    ) -> Self {
+        let mut sensors = HashMap::new();
+        let mut by_component = HashMap::new();
+        for cfg in cfgs {
+            if cfg.sensors.is_empty() {
+                continue;
+            }
+            if !cfg
+                .capabilities
+                .contains(DeviceCapabilities::HAS_MEASUREMENT_CHANNELS)
+            {
+                panic!(
+                    "invalid component config: a device with sensors should \
+                     have the `HAS_MEASUREMENT_CHANNELS` capability:{cfg:#?}"
+                );
+            }
+
+            let mut ids = Vec::with_capacity(cfg.sensors.len());
+            for sensor in &cfg.sensors {
+                let sensor_id = sensor.def.sensor_id;
+                let prev = sensors.insert(
+                    sensor_id,
+                    Sensor {
+                        def: sensor.def.clone(),
+                        state: sensor.state.clone(),
+                    },
+                );
+                assert!(
+                    prev.is_none(),
+                    "invalid config: sensor ID {sensor_id} already exists!\n
+                     previous sensor: {prev:#?}\nnew sensor: {sensor:#?}"
+                );
+                ids.push(sensor_id)
+            }
+
+            let component = SpComponent::try_from(cfg.id.as_str()).unwrap();
+            let prev = by_component.insert(component, ids);
+            assert!(prev.is_none(), "component ID collision!");
+        }
+        Self { sensors, by_component }
+    }
+
+    fn sensor_for_component<'sensors>(
+        &'sensors self,
+        component: &SpComponent,
+        index: BoundsChecked,
+    ) -> Option<&'sensors Sensor> {
+        let id = self.by_component.get(component)?.get(index.0 as usize)?;
+        self.sensors.get(id)
+    }
+
+    pub(crate) fn num_component_details(
+        &self,
+        component: &SpComponent,
+    ) -> Option<u32> {
+        let len = self
+            .by_component
+            .get(component)?
```

Comment:
So, the `component_details` construct is used for both components which have "measurement channels" (e.g. sensors) _and_ for switch ports, which have a `PortStatus` details type. The simulated sidecar SP doesn't presently implement the `PortStatus` details type, but I figured that it was better for this function to return an `Option` so that the simulated sidecar could, eventually, know to also check in the list of port statuses, rather than panicking.

I could just make it panic for now, but presently, it's the caller who panics in this case when it gets a `None`, so that the sidecar impl could eventually choose not to panic.

Reply:
All right, I think you know more about this than I do so your judgment!

Reply:
I'm very much open to being convinced otherwise :woman_shrugging:

Reply:
I think neither this nor the caller should panic - the `component` coming in here is from an external request, presumably via MGS or faux-mgs, neither of which knows whether a given component name exists on the target SP until it asks (which eventually lands in this method). E.g.,

```
john@jeeves ~ $ pilot -r madrid sp exec -e 'component-details foobar' BRM42220004
Aug 19 18:22:49.330 INFO creating SP handle on interface madrid_sw0tp0, component: faux-mgs
Aug 19 18:22:49.360 INFO initial discovery complete, addr: [fe80::aa40:25ff:fe04:147%15]:11111, interface: madrid_sw0tp0, component: faux-mgs
Error: Error response from SP: unsupported request for this SP component
```

Maaaybe the error there should be more specific ("this SP component" implies it's a real component when the actual error is "`foobar` isn't a valid component name for this SP"), but that's an aside to this question - we should certainly accept arbitrary component names and not panic.

Reply:
@jgallagher ahh, I was imagining that validation being done in a higher later such that this code gets to assume validity.